akka-stream/src/main/scala/akka/stream/scaladsl/Source.scala

```diff
@@ -437,6 +437,18 @@ object Source {
   def lazilyAsync[T](create: () ⇒ Future[T]): Source[T, Future[NotUsed]] =
     lazily(() ⇒ fromFuture(create()))
 
+  /**
+   * Creates a real `Source` upon receiving the first demand. Internal `Source` will not be created if downstream
+   * finishes without any demand.
+   *
+   * If downstream finishes without any demand the materialized `Future` is completed with `None`.
+   * If the future of the real source or the materialization of the real source fails then the
+   * materialized `Future` is completed with the exception.
+   * Otherwise the `Future` is completed with `Some` of the materialized value of the internal source.
+   */
+  def lazyInitAsync[T, M](sourceFactory: () ⇒ Future[Source[T, M]]): Source[T, Future[Option[M]]] =
+    Source.fromGraph(new LazySourceAsync[T, M](sourceFactory))
+
```

Comment:
Unfortunate with the name clash here, since there is a `lazily(() => Source)` one would have wanted it to be `lazilyAsync` which instead would have been better named `lazilySingle` (since there is `Source.single`). 

And the semantics difference where `lazily` fails the returned future if there is no subscription while this completes the future with `None`.

Reply:
We can't change those names though because of binary and source compatibility, the best we can do is to deprecate things.